haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,211 @@
+        :type batch_size: int (optional)
+        :return: A list of dictionaries, where each dictionary contains the question, positive document,
+                and negative document.
+        """
+        question_pos_doc_neg_doc: List[Dict[str, str]] = []
+
+        for i in range(0, len(question_doc_pairs), batch_size):
+            # question in batches to minimize network latency
+            i_end = min(i + batch_size, len(question_doc_pairs))
+            queries: List[str] = [e["question"] for e in question_doc_pairs[i:i_end]]
+            pos_docs: List[str] = [e["document"] for e in question_doc_pairs[i:i_end]]
+
+            docs: List[List[Document]] = self.retriever.retrieve_batch(
+                queries=queries, top_k=self.top_k, batch_size=batch_size
+            )
+
+            # iterate through queries and find negatives
+            for question, pos_doc, top_docs in zip(queries, pos_docs, docs):
+                random.shuffle(top_docs)
+                for doc_item in top_docs:
+                    neg_doc = doc_item.content
+                    if neg_doc != pos_doc:
+                        question_pos_doc_neg_doc.append({"question": question, "pos_doc": pos_doc, "neg_doc": neg_doc})
+                        break
+        return question_pos_doc_neg_doc
+
+    def generate_margin_scores(self, mined_negatives: List[Dict[str, str]]) -> List[Dict]:
+        """
+        Given a list of mined negatives, predict the score margin between the positive and negative document using
+        the cross encoder.
+
+        The function returns a list of examples, where each example is a dictionary with the following keys:
+
+        * question: the question string
+        * pos_doc: the positive document string
+        * neg_doc: the negative document string
+        * score: the score margin
+
+        :param mined_negatives: List[Dict[str, str]] = List of mined negatives
+        :type mined_negatives: List[Dict[str, str]]
+        :return: A list of dictionaries, each of which has the following keys:
+            - question: The question string
+            - pos_doc: The positive document string
+            - neg_doc: The negative document string
+            - score: The score margin
+        """
+        examples: List[Dict] = []
+        for item in mined_negatives:
+            scores = self.cross_encoder.predict(
+                [[item["question"], item["pos_doc"]], [item["question"], item["neg_doc"]]]
+            )
+            score_margin = scores[0] - scores[1]
+            examples.append(
+                {
+                    "question": item["question"],
+                    "pos_doc": item["pos_doc"],
+                    "neg_doc": item["neg_doc"],
+                    "score": score_margin,
+                }
+            )
+        return examples
+
+    def generate_pseudo_labels(self, documents: List[Document]) -> Tuple[dict, str]:
```

Comment:
Here, we can also allow a batch_size parameter that will be passed to the called methods.